lib/hsa/mcwamp_hsa.cpp

```diff
@@ -2260,64 +2260,74 @@ class HSADevice final : public KalmarDevice
+        }
 
-            // Allocate a new queue, we are belowthe HCC_MAX_QUEUES limit :
-            //
+        this->rocrQueuesMutex.unlock();
 
-            auto rq = new RocrQueue(agent, this->queue_size, thief);
-            rocrQueues.push_back(rq);
+        if (foundRQ != nullptr)
+            return;
 
-            DBOUT(DB_QUEUE, "Create new rocrQueue=" << rq << " for thief=" << thief << "\n")
+        // Steal an unused queue when we reaches the limit
+        while (!foundRQ) {
 
-        } else {
-            RocrQueue *foundRQ = nullptr;
-            while (!foundRQ) {
-                // First make a pass to see if we can find an unused queue:
-                for (auto rq : rocrQueues) {
-                    if (rq->_hccQueue == nullptr) {
-                        DBOUT(DB_QUEUE, "Found unused rocrQueue=" << rq << " for thief=" << thief << ".  hwQueue=" << rq->_hwQueue << "\n")
-                        foundRQ = rq;
-                        break;
-                    }
-                }
-                if (!foundRQ) {
-                    for (auto rq : rocrQueues) {
-                        if (rq->_hccQueue != thief)  {
-                            auto victimHccQueue = rq->_hccQueue;
-                            // victimHccQueue==nullptr should be detected by above loop.
-                            std::lock_guard<std::mutex> (victimHccQueue->qmutex);
-                            if (victimHccQueue->isEmpty()) {
-                                DBOUT(DB_LOCK, " ptr:" << this << " lock_guard...\n");
-
-                                assert (victimHccQueue->rocrQueue == rq);  // ensure the link is consistent.
-                                victimHccQueue->rocrQueue = nullptr;
-                                foundRQ = rq;
-                                DBOUT(DB_QUEUE, "Stole existing rocrQueue=" << rq << " from victimHccQueue=" << victimHccQueue << " to hccQueue=" << thief << "\n")
-                                break;
-                            }
-                        }
-                    }
+            this->rocrQueuesMutex.lock();
+
+            // First make a pass to see if we can find an unused queue
+            for (auto rq : rocrQueues) {
+                if (rq->_hccQueue == nullptr) {
+                    DBOUT(DB_QUEUE, "Found unused rocrQueue=" << rq << " for thief=" << thief << ".  hwQueue=" << rq->_hwQueue << "\n")
+                    foundRQ = rq;
+                    break;
                 }
-                if (foundRQ) {
+            }
 
-                    // update the queue pointers to indicate the theft:
-                    foundRQ->assignHccQueue(thief);
+            this->rocrQueuesMutex.unlock();
 
-                    break; // while !foundVictim
-                };
+            if (foundRQ != nullptr) {
+                // update the queue pointers to indicate the theft
+                foundRQ->assignHccQueue(thief);
```

Comment:
should you do this right after this (https://github.com/RadeonOpenCompute/hcc/pull/624/files#diff-410c18478a22065083790061957b4770R2290) while you still have the lock; otherwise it looks like there will be a race condition

Reply:
thx. I'll change it right away.